chain-signatures/node/src/protocol/signature.rs

```diff
@@ -87,42 +89,67 @@ impl SignQueue {
             }
         } {
             let mut rng = StdRng::from_seed(request.entropy);
-            let subset = stable.keys().choose_multiple(&mut rng, threshold);
-            let proposer = **subset.choose(&mut rng).unwrap();
-            if subset.contains(&&self.me) {
-                let is_mine = proposer == self.me;
+            let subset = stable.keys().cloned().choose_multiple(&mut rng, threshold);
+            let in_subset = subset.contains(&self.me);
+            let proposer = *subset.choose(&mut rng).unwrap();
+            let is_mine = proposer == self.me;
+            let request_id = CryptoHash(request.request_id);
+
+            tracing::info!(
+                ?request_id,
+                ?subset,
+                ?proposer,
+                in_subset,
+                is_mine,
+                "sign queue: organizing request"
+            );
+
+            let sign_id = SignRequestIdentifier::new(
+                request.request_id,
+                request.epsilon,
+                request.request.payload,
+            );
+
+            if in_subset {
                 tracing::info!(
-                    request_id = ?CryptoHash(request.request_id),
-                    ?is_mine,
-                    ?subset,
-                    ?proposer,
+                    ?request_id,
                     "saving sign request: node is in the signer subset"
                 );
-                let proposer_requests = self.requests.entry(proposer).or_default();
-                proposer_requests.push_back(request);
+
                 if is_mine {
                     crate::metrics::NUM_SIGN_REQUESTS_MINE
                         .with_label_values(&[my_account_id.as_str()])
                         .inc();
+                    self.my_requests.push_back((request, proposer, subset));
+                } else {
+                    self.other_requests
+                        .insert(sign_id, (request, proposer, subset));
                 }
             } else {
                 tracing::info!(
-                    rrequest_id = ?CryptoHash(request.request_id),
-                    me = ?self.me,
-                    ?subset,
-                    ?proposer,
+                    ?request_id,
                     "skipping sign request: node is NOT in the signer subset"
                 );
             }
         }
     }
 
-    pub fn take_my_requests(&mut self) -> VecDeque<SignRequest> {
-        self.requests.remove(&self.me).unwrap_or_default()
+    pub fn take_my_requests(&mut self) -> VecDeque<(SignRequest, Participant, Vec<Participant>)> {
+        std::mem::take(&mut self.my_requests)
```

Comment:
Is it safe to work with memory here?
Does it affect performance? Is it a blocking operation for incoming requests?

Reply:
the only time where anything is unsafe is if the code shows `unsafe`. For the most part, any time you see `std::mem` it's an optimization on accessing the heap. This isn't an IO operation so not blocking and purely just multiple assignment operations